Supports converting Positions and Trades to pandas dataframes.

bryanehansen opened the pull request:
Fixes #6. Add convenient Pandas conversions

bryanehansen committed:
Supports converting Positions and Trades to pandas dataframes.

Fixes #6. Add convenient Pandas conversions

bryanehansen commented on bankroll/model/modelconverter.py:
```diff
@@ -0,0 +1,16 @@
+from bankroll.model import ConvertableModel, Position, Trade
+
+
+class ModelConverter:
```
I'd like to find a better way to solve this than adding the ModelConverter class. Ideas?

jspahrsummers replied:
Why does this need a class at all? Can we use a global function?

bryanehansen replied:
✅

bryanehansen commented on bankroll/model/position.py:
```diff
@@ -44,6 +45,13 @@ def __post_init__(self) -> None:
 
+    @staticmethod
+    def dataframeColumns() -> List[str]:
+        return ["Instrument", "Quantity", "Avg Price"]
```
Ideally, I'd use a tuple to relate the function call to obtain the value to the header column it should go under as a private mapping. This would allow the order of headers/values to be group correctly.

jspahrsummers replied:
What didn't work when you tried that?

bryanehansen replied:
I wasn't sure how it would work with all of the following functions to obtain values:
```
        return [
            self.date.date(),
            self.action,
            abs(self.quantity),
            self.instrument,
            self.amount,
            self.fees,
        ]
```
…specifically `abs(self.quantity)`

jspahrsummers replied:
They're all of kind `Position -> Any`, so the functions have uniform type. They'll probably need to be defined with lambdas.

bryanehansen replied:
✅ Got this sorted out nicely

bryanehansen commented on bankroll/model/convertablemodel.py:
```diff
@@ -0,0 +1,13 @@
+from typing import Any, List
+
+
+class ConvertableModel(ABC):
```
I'm still having typing issues with this approach.

jspahrsummers replied:
Can you be more specific? What sort of issues?

jspahrsummers replied:
Nit: can we name this `PandasConvertibleModel` (note the `i`) or something like that?

bryanehansen replied:
✅  Figured out the typing issues

jspahrsummers commented on bankroll/model/activity.py:
```diff
@@ -78,11 +79,30 @@ def price(self) -> Cash:
+        return action
+
+    @staticmethod
+    def dataframeColumns() -> List[str]:
```
I'm not a big fan of this living within the main model definition, if it sort of adds bloat without benefit to non-Pandas users. Can it be defined externally to the class in a sane way?

bryanehansen replied:
I extracted this out, although the `dataframeColumnFunctions` now will have an if/elif sequence for all model types that need to be converted

jspahrsummers commented on bankroll/model/activity.py:
```diff
@@ -78,11 +79,30 @@ def price(self) -> Cash:
+            self.instrument,
+            self.amount,
+            self.fees,
+        ]
```
What you're doing here is basically the same as `dataclasses.asdict`. The issue I think you're going to run into (with that function and this approach) is that the nested models (like `instrument`) should ideally be unpacked before `DataFrame` conversion, or else we've stuck a complex Python object into a `DataFrame`, which will make it really challenging to manipulate.

bryanehansen replied:
Currently I am ensuring that the `instrument` uses its string representation. This seems sufficient at the moment to get identical output to the activity command for the csv output. Can we revisit this once we need a nested representation?

jspahrsummers commented on bankroll/model/modelconverter.py:
```diff
@@ -0,0 +1,16 @@
+import pandas as pd
```
This should be an optional dependency, so we don't crash anyone who doesn't have it installed.

bryanehansen replied:
✅ pandas are now optional 🐼

jspahrsummers commented on bankroll/model/modelconverter.py:
```diff
@@ -0,0 +1,16 @@
+    def dataframe(cls, items: List[ConvertableModel]) -> pd.DataFrame:
+        if len(items) > 0:
+            firstItem = items[0]
+            assert all(isinstance(i, type(firstItem)) for i in items)
```
I understand what this check is doing, but why is it needed?

bryanehansen replied:
The columns only make sense for the specific Trade or Position classes, not both.

jspahrsummers commented on bankroll/model/modelconverter.py:
```diff
@@ -0,0 +1,16 @@
+        if len(items) > 0:
+            firstItem = items[0]
+            assert all(isinstance(i, type(firstItem)) for i in items)
+            rows = list(map(lambda t: t.dataframeValues(), items))
```
I suggest a list comprehension instead:

```py
rows = [t.dataframeValues() for t in items]
```

jspahrsummers commented on bankroll/converter/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from .modelconverter import ModelConverter
+
+# __all__ = ["ModelConverter"]
```
Did you mean to comment this out?

bryanehansen replied:
✅ Now using the converter package properly

jspahrsummers reviewed (changes requested):
Thanks for taking a stab at this!

If you want to proceed with the `converter` package, you'll also need to add it as one of the packages installed by `setup.py` (see where `bankroll.model` is specified in there, and do something similar).

bryanehansen committed:
Fixed abstract class typing issues. Moved converter to bankroll.converter package.

bryanehansen committed:
Make pandas an optional package

bryanehansen committed:
Extract pandas conversion logic from model classes.

jspahrsummers commented on bankroll/converter/modelconverter.py:
```diff
@@ -0,0 +1,74 @@
+from typing import Any, Callable, cast, List, Optional, Sequence, Tuple, Union
+
+try:
+    import pandas as pd
```
To fix the [build failure](https://circleci.com/gh/bankroll-py/bankroll-model/19):
```suggestion
    import pandas as pd # type: ignore
```

ShadoFlameX replied:
✅

jspahrsummers commented on bankroll/converter/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from .modelconverter import dataframeForModelObjects
```
Just speaking structurally, I feel like it's a bit weird to have a package named `bankroll.converter.modelconverter`, when we could have `bankroll.model.converter`, y'know?

If you want to make sure that the imported functions are namespaced like "converter" functions, you could structure the imports as so:

```py
import bankroll.model.converter as converter

converter.dataframeForModelObjects(...)
```

You could also make sure that `bankroll/model/__init__.py` only exports `converter` the module and not anything inside, to prevent people using the "naked" symbols by default.

jspahrsummers commented on bankroll/converter/modelconverter.py:
```diff
@@ -0,0 +1,74 @@
+try:
+    import pandas as pd
+except ImportError:
+    pass
```
IIRC this still needs to set `pandas` to avoid an undefined symbol error if the name is ever referenced:
```suggestion
    pandas = None
```

ShadoFlameX replied:
✅

jspahrsummers commented on bankroll/converter/modelconverter.py:
```diff
@@ -0,0 +1,74 @@
+from bankroll.model import Position, Trade
+
+ConvertibleModel = Union[Position, Trade]
+ColumnFunctions = List[Tuple[str, Callable[[Any], Any]]]
```
Python 3 dictionaries actually preserve insertion order (which I believe is the reason you used a `List` here), so this can actually be a `Dict`, which is much simpler.

```suggestion
_ColumnFunctions = Dict[str, Callable[[_ConvertibleModel], Any]]
```

ShadoFlameX replied:
✅

jspahrsummers commented on bankroll/converter/modelconverter.py:
```diff
@@ -0,0 +1,74 @@
+
+def dataframeForModelObjects(
+    items: Sequence[ConvertibleModel]
+) -> Optional[pd.DataFrame]:
```
I don't think this should return `Optional`... it should probably throw an exception if `pandas` isn't available, so errors aren't silently missed.
```suggestion
) -> pd.DataFrame:
```

ShadoFlameX replied:
✅

jspahrsummers commented on bankroll/converter/modelconverter.py:
```diff
@@ -0,0 +1,74 @@
+) -> Optional[pd.DataFrame]:
+    df = None
+
+    if module_exists("pandas"):
```
If you set `pandas = None` like I suggested above, this can just be written as:

```suggestion
    if pandas:
```

jspahrsummers commented on bankroll/converter/modelconverter.py:
```diff
@@ -0,0 +1,74 @@
+
+from bankroll.model import Position, Trade
+
+ConvertibleModel = Union[Position, Trade]
```
This is a good way of doing it! It allows collections of heterogeneous values, though, as you've seen (and guarded against).

You could instead use a `TypeVar` to statically enforce that all of the values are the same type:

```suggestion
_ConvertibleModel = TypeVar('_ConvertibleModel', Position, Trade)
```

(sidenote: please prefix all private symbols with `_`)

jspahrsummers commented on bankroll/converter/modelconverter.py:
```diff
@@ -0,0 +1,74 @@
+            else:
+                assert type(firstItem) == type(i)
+
+            rows.append([t[1](i) for t in dataframeColumnFunctions(i)])
```
With the change to using a `Dict` and a `TypeVar` as mentioned above, this whole loop could be collapsed into:

```suggestion
            rows = [[fn(i) for fn in dataframeColumnFunctions.values()] for i in items]
```

jspahrsummers commented on bankroll/converter/modelconverter.py:
```diff
@@ -0,0 +1,74 @@
+
+            rows.append([t[1](i) for t in dataframeColumnFunctions(i)])
+
+        if firstItem is not None:
```
This can be omitted with the `TypeVar` solution above.

jspahrsummers commented on bankroll/converter/modelconverter.py:
```diff
@@ -0,0 +1,74 @@
+
+def dataframeColumnFunctions(model: ConvertibleModel) -> ColumnFunctions:
+    if isinstance(model, Position):
+        posInstrument: Callable[[Position], Any] = lambda p: str(p.instrument)
```
I'm guessing you wrote all these temporary variables to make the types unify.

I _think_ you should be able to achieve the same thing by just saving the dictionary into a local:

```py
cf: ColumnFunctions = {
  'Instrument': lambda p: str(p.instrument),
  'Quantity': lambda p: p.quantity.normalize()
  ...
}

return cf
```

jspahrsummers commented on bankroll/converter/modelconverter.py:
```diff
@@ -0,0 +1,74 @@
+    return []
+
+
+def module_exists(module_name: str) -> bool:
```
This can be removed with my `pandas = None` suggestion.

bryanehansen replied:
✅

jspahrsummers commented on bankroll/converter/modelconverter.py:
```diff
@@ -0,0 +1,74 @@
+
+
+def dataframeForModelObjects(
+    items: Sequence[ConvertibleModel]
```
I _think_ everything you're doing is covered by the `Iterable` interface (since you don't iterate the collection twice):

```suggestion
    items: Iterable[ConvertibleModel]
```

This is a bit more general, and supports things like generator comprehensions: `dataframeForModelObjects((x.z for x in y))`

ShadoFlameX replied:
Actually I do iterate twice, which is why I changed it to a `Sequence`

jspahrsummers commented on tests/test_converter.py:
```diff
@@ -0,0 +1,39 @@
+
+
+class TestModelConverter(unittest.TestCase):
+    @given(lists(from_type(Trade), min_size=3, max_size=3))
```
I would recommend `min_size=0` for maximum testing, unless you have a specific reason why not.

```suggestion
    @given(lists(from_type(Trade), max_size=3))
```

jspahrsummers commented on tests/test_converter.py:
```diff
@@ -0,0 +1,39 @@
+            self.assertEqual(df.at[i, "Amount"], trades[i].amount)
+            self.assertEqual(df.at[i, "Fees"], trades[i].fees)
+
+    @given(lists(from_type(Position), min_size=3, max_size=3))
```
As above.

jspahrsummers reviewed (changes requested):
Great tests, thank you! This is looking pretty good. 😎 

Most of my comments are cosmetic or nice-to-haves, so do what you will with those, but let's definitely resolve the module/package naming discussion before shipping this.

bryanehansen committed:
Cleanup after review.

bryanehansen committed:
Allow any Activity for dataframeForModelObjects. Changing converter packaging.

jspahrsummers committed:
Add pandas to requirements.txt

jspahrsummers committed:
Move `converter` module into `model` package

jspahrsummers committed:
Fix type error in _ColumnFunctions

And `t` has to be cast because _technically_ the returned Callable is allowed to be invoked with any `Activity` (for complex type unification rules that aren't worth going into... but TLDR, mypy is right to warn if you remove the casts).

jspahrsummers committed:
Fix type error on pandas import in test

jspahrsummers committed:
Use Sequence instead of List

Sequence is covariant: https://mypy.readthedocs.io/en/latest/common_issues.html#variance

jspahrsummers committed:
Add CI step to run typechecking without pandas available

jspahrsummers committed:
Raise error in dataframeForModelObjects if Pandas is missing

jspahrsummers committed:
Prefix dataframeColumnFunctions with _ to indicate it's private

bryanehansen commented:
👍 Thanks for the help here, this looks good to me!
